### PR TITLE
Add Pie Chart Type to UI

### LINKED ISSUE #873: Add Pie Chart Type to UI
We can already show these types of charts/graphs:
- Network map
- Timeseries (multiple stacked line graphs)
- Bar charts
- Histograms
- Flamecharts (`px/perf_flamegraph`, for example)
- Basic stat charts
- Basic text charts

We could also use a pie chart, representing the same data as the bar chart does.

It could look something like this:
image

### PR DESCRIPTION
**Pull request by NickLanam**
Summary: Adds a pie chart. Give it one column to group/label with, and another to sum values from, and it'll output something like this.
image
Fixes #873

Relevant Issues: #873

Test Plan: Run this script with this vis spec.
Try hovering the individual pie slices. Compare the results to what's in the data drawer.
Try resizing the window so that the pie chart resizes to fit its container.
Try running the script more than once, to see how the color assignment behaves.
It should never be able to put two slices of the same color next to each other.
```python
import px
import pxviews

ns_per_ms = 1000 * 1000
ns_per_s = 1000 * ns_per_ms
# Window size to use on time_ column for bucketing.
window_ns = px.DurationNanos(10 * ns_per_s)

def process_stats_by_namespace(start_time: str):
    ''' Gets a summary of process stats by namespace since `start_time`.
        Computes the total I/O consumption across the namespace since `start_time`.

    Args:
    @start_time Start time of the data to examine.
    '''
    df = pxviews.pod_resource_stats(start_time, px.now())
    df = df.groupby('namespace').agg(
        vsize=('vsize', px.sum),
        rss=('rss', px.sum),
        actual_disk_read_throughput=('actual_disk_read_throughput', px.sum),
        actual_disk_write_throughput=('actual_disk_write_throughput', px.sum),
        total_disk_read_throughput=('total_disk_read_throughput', px.sum),
        total_disk_write_throughput=('total_disk_write_throughput', px.sum),
    )

    return df[[
        'namespace',
        'vsize',
        'rss',
        'actual_disk_read_throughput',
        'actual_disk_write_throughput',
        'total_disk_read_throughput',
        'total_disk_write_throughput',
    ]]
```
```json
{
    "variables": [
        {
            "name": "start_time",
            "type": "PX_STRING",
            "description": "The relative start time of the window. Current time is assumed to be now",
            "defaultValue": "-5m"
        }
    ],
    "widgets": [
        {
            "name": "VSize by Namespace",
            "position": {
                "x": 0,
                "y": 0,
                "w": 3,
                "h": 2
            },
            "func": {
                "name": "process_stats_by_namespace",
                "args": [
                    {
                        "name": "start_time",
                        "variable": "start_time"
                    }
                ]
            },
            "displaySpec": {
                "@type": "types.px.dev/px.vispb.PieChart",
                "pie": {
                    "value": "vsize",
                    "label": "namespace"
                }
            }
        }
    ],
    "globalFuncs": []
}
```

Type of change: /kind feature

Changelog Message:
```release-note
New chart type: pie
```

Signed-off-by: Nick Lanam <nlanam@pixielabs.ai>

### CODE REVIEW

**Comment by NickLanam** [comment]
Minor note: the color palette given to Vega here isn't quite what it used. Vega's docs don't seem to mention this, giving it a `scheme` of `['#FF0000', '#00FF00', '#0000FF']` ends up using every shade of red, green, and blue _except_ the ones given, and it even puts a brown in there.

**Comment by JamesMBartlett** [comment]
> This also holds if giving it one of the [predefined color schemes](https://vega.github.io/vega/docs/schemes/): it uses them as an inspiration, instead of using them directly.
It seems like using their builtin color schemes it does the right thing without interpolation: https://github.com/vega/vega/blob/7dbec7ddd0e94c8150e0f90857d851c72776ad0e/packages/vega-encode/src/Scale.js#L317-L323

We could just register our color scheme with vega with https://vega.github.io/vega/docs/api/extensibility/#scheme

**Comment by NickLanam** [comment]
> 

It seems that isn't the code path it ends up using (at least, that isn't the entire story). When I gave it the builtin scheme `category20`, for instance, it used a combination of the colors in that theme and a few that it interpolated from there. It didn't need to do that: there are 20 colors in that scheme, but only eight slices to colorize in that test. Whatever it's doing, it's more complex than it looks.

**Comment by NickLanam** [comment]
> When I look at `category20` [here](https://vega.github.io/editor/#/url/vega/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykSArJQBWENgDsQAGhAATONABONHJnaT0AQQAETJBBpRtOGnG2wkizNrgAPJAhwAbOJWkgA7jVn00AJgAGAJl4GjIsfyCZJAZMNgMALzg0EHEJZJkDMnEkJwg0AG1QHIRk9AhMS0xNcTIXd2JchjLgkCYacVk0UA6cWJTFJFqMkAQOtFbkWzQANko-AE5MzDgcCcoAgEYAX22pYocykDhOmrqRxqdm2fmlto6u1B7xPswBobIRsY1JpGnUOaLZardZbXb7VKHFI4JCyM71GSXa6oVrtTrdEC9froQbDdzfCYyKbrHZ7A6lFIdcRwRQAJVhNAY+URTRaMjRj2er3eeKJ4xRRL+aAWrQqINQpIhJSOUDYimpdIZTIarMJ93RT0xL2xIFxn3x-N+-02opWaxRlAE4PJRzEVhVVzKADNchA4OyHhisW90LA4FAANZMNjTXYAXRksiQlUKNpSlSYCJASIUscxj02MidpicjwALGT0-4sznHjNC940ABmEtwXNoE0V-O1+uoKtNtACFuPADsHYB3bQAA5thGQJhcRAnXKEGnMABPHBHEwjbN1x4gNf15ZVeFlUBZHJOFIVXe1ep7Y6nc-7kCH3IpE5wm8gS92n0HsJHk9yt7h0eZFAuSpqgRSQhSvpsE4cruAuS4pHKsgdA+kZsMg-KgFGMboAmSZbhulaXnqt5cIgMrRuQcrzoEr4AaMlgBvkoGgHBRyWFA7hOooaEYlhKA4UgibJJeJyyvIGInCsigYtmTjHpq0DASksrQdJg7oIRMj-J+2QPugXg+DA2gENofivjI84Yve8kYHAYRYMZpm0TIDA4FhJGVFYe4yaWJ6edUL4ide5w+euj7BReMgws+IUKV+ekgNF3mXlSNL0khypxbpNmpYqGX5JebCxGlSpMTp376d49COWZl6yvKJX5VZ8U2fVCrpYyBW7KO2xAA) It only uses the colors in the color scheme and doesn't interpolate

Closing the loop: it seems that the demo does as you say - where it only uses the colors in the theme if it's a named one - but something about our setup makes it interpolate anyway.

Registering a _custom_ theme, however, does use only the given colors. That isn't ideal either, though, because it then sometimes gives the same color to two neighboring slices:
image

I could probably fix that, but the `hsl`-interpolated colors already do the right thing and look okay. Perhaps it's best to let it interpolate.

**Review by JamesMBartlett — APPROVED** [review]

File: src/ui/src/containers/live/canvas.tsx [hunk 1]
```diff
@@ -134,6 +134,7 @@ const useStyles = makeStyles((theme: Theme) => createStyles({
     '&.focus': {
       border: `1px solid ${theme.palette.foreground.grey2}`,
     },
+    minHeight: 0,
```
[the remark this answers is not in the document]
Comment (NickLanam) [reply src/ui/src/containers/live/canvas.tsx:137]: Yes. Without it, the pie chart (as a flex item) does not correctly compute its container's height and will grow indefinitely beyond it. This is a very common situation for flex children, we use the same fix elsewhere too.

**Pull request merged**